### PR TITLE
Add guest_vlan_allowed property to routed and isolated VCD networks

### PR DESCRIPTION
**Pull request by s-winkelhofer**
This PR enables the `guest_vlan_allowed` property on routed and isolated VCD networks (feature available on VCD 10.3.0+)

### CODE REVIEW

File: vcd/resource_vcd_network_isolated_v2_nsxt_test.go [hunk 1]
```diff
@@ -122,33 +125,35 @@ resource "vcd_network_isolated_v2" "net1" {
...
-  metadata = {
-    {{.MetadataKey}} = "{{.MetadataValue}}"
-  }
+	  start_address = "1.1.1.10"
+	  end_address   = "1.1.1.20"
+	}
+	
+	metadata = {
+		{{.MetadataKey}} = "{{.MetadataValue}}"
+	}
```
Comment (adezxc) [inline vcd/resource_vcd_network_isolated_v2_nsxt_test.go:137]: For consistency we keep these configuration templates in line with Terraform style guidelines
```suggestion
  start_address = "1.1.1.10"
  end_address   = "1.1.1.20"
}
	
metadata = {
  {{.MetadataKey}} = "{{.MetadataValue}}"
}
```

File: vcd/resource_vcd_network_isolated_v2_nsxt_test.go [hunk 2]
```diff
@@ -122,33 +125,35 @@ resource "vcd_network_isolated_v2" "net1" {
...
-  org  = "{{.Org}}"
-  vdc  = "{{.NsxtVdc}}"
-  
-  name        = "{{.NetworkName}}"
-  description = "updated NSX-T isolated network test"
+	org  = "{{.Org}}"
+	vdc  = "{{.NsxtVdc}}"
+	
+	name        = "{{.NetworkName}}"
+	description = "updated NSX-T isolated network test"
```
Comment (adezxc) [inline vcd/resource_vcd_network_isolated_v2_nsxt_test.go:147]: For consistency we keep these configuration templates in line with Terraform style guidelines
```suggestion
  org  = "{{.Org}}"
  vdc  = "{{.NsxtVdc}}"
	
  name        = "{{.NetworkName}}"
  description = "updated NSX-T isolated network test"
```

File: vcd/resource_vcd_network_routed_v2_nsxt_test.go [hunk 3]
```diff
@@ -170,6 +175,7 @@ resource "vcd_network_routed_v2" "net1" {
 
   gateway = "1.1.1.1"
   prefix_length = 24
+  guest_vlan_allowed = true
```
Comment (adezxc) [inline vcd/resource_vcd_network_routed_v2_nsxt_test.go:178]: For consistency we keep these configuration templates in line with Terraform style guidelines
```suggestion
  gateway                    = "1.1.1.1"
  prefix_length            = 24
  guest_vlan_allowed = true
```

File: vcd/resource_vcd_network_isolated_v2_nsxt_test.go [hunk 4]
```diff
@@ -122,33 +125,35 @@ resource "vcd_network_isolated_v2" "net1" {
...
-
+  guest_vlan_allowed = true
+  
   static_ip_pool {
-	start_address = "1.1.1.10"
-	end_address   = "1.1.1.20"
-  }
+	  start_address = "1.1.1.10"
+	  end_address   = "1.1.1.20"
+	}
```
Comment (adezxc) [inline vcd/resource_vcd_network_isolated_v2_nsxt_test.go:156]: For consistency we keep these configuration templates in line with Terraform style guidelines
```suggestion
  start_address = "1.1.1.10"
  end_address   = "1.1.1.20"
}
```

**Review by adezxc — CHANGES_REQUESTED** [review]
Hi @s-winkelhofer, thanks for your contribution!

A few comments: 
* Regarding @vmwclabot, you need to sign your commits with `Signed-off-by: John Doe <john.doe@email.org>` at the end of the commit message, you can also do that with the `-s` flag.
* We want to mention every feature/improvement/bugfix PR in the changelog, for that you should create a file in `.changes/v3.10.0`, see [here] for reference
* In this repo we also have the documentation for the Terraform registry, so you should document the new argument in all the resources/data sources that you changed, and add `guest_vlan_allowed` to the example configuration section.


If you have any questions, please don't hesitate to ask us how to proceed, thanks again!

[here]: https://github.com/vmware/terraform-provider-vcd/blob/main/.changes/v3.9.0/1010-improvements.md

File: vcd/datasource_vcd_network_routed_v2.go [hunk 5]
```diff
@@ -102,6 +103,11 @@ func datasourceVcdNetworkRoutedV2() *schema.Resource {
 				Description: "IP ranges used for static pool allocation in the network",
 				Elem:        networkV2IpRangeComputed,
 			},
+			"guest_vlan_allowed": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (Didainius) [inline vcd/datasource_vcd_network_routed_v2.go:108]:
```suggestion
				Computed:    true,
```

File: vcd/datasource_vcd_network_routed.go [hunk 6]
```diff
@@ -89,6 +90,12 @@ func datasourceVcdNetworkRouted() *schema.Resource {
 				Description: "Defines if this network is shared between multiple VDCs in the Org",
 			},
 
+			"guest_vlan_allowed": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (Didainius) [inline vcd/datasource_vcd_network_routed.go:95]:
```suggestion
				Computed:    true,
```

File: vcd/datasource_vcd_network_isolated_v2.go [hunk 7]
```diff
@@ -98,6 +98,11 @@ func datasourceVcdNetworkIsolatedV2() *schema.Resource {
 				Description: "IP ranges used for static pool allocation in the network",
 				Elem:        networkV2IpRangeComputed,
 			},
+			"guest_vlan_allowed": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (Didainius) [inline vcd/datasource_vcd_network_isolated_v2.go:103]:
```suggestion
				Computed:    true,
```

File: vcd/datasource_vcd_network_isolated.go [hunk 8]
```diff
@@ -132,6 +133,11 @@ func datasourceVcdNetworkIsolated() *schema.Resource {
 					},
 				},
 			},
+			"guest_vlan_allowed": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (Didainius) [inline vcd/datasource_vcd_network_isolated.go:138]:
```suggestion
				Computed:    true,
```

**Comment by s-winkelhofer** [comment]
Thanks for the review, I updated the branch, so it covers all your suggested changes.

**Review by adezxc — APPROVED** [review]
Looks good to me, tested the change manually, thanks for the PR!

**Comment by Didainius** [comment]
@s-winkelhofer ,
Thanks for the adjustments! I wanted to clarify a general thing as it is valuable feedback. The `_v2` resources are made for the replacement of previous resources (non `_v2`) so I am curious - are you still actively using any of the `vcd_network_routed` or `vcd_network_isolated`?

**Comment by s-winkelhofer** [comment]
@Didainius thank you. We are only using the `_v2` resources. I added the feature to the `non _v2` resources for completeness reasons.

File: website/docs/r/network_isolated.html.markdown [hunk 9]
```diff
@@ -60,6 +62,7 @@ The following arguments are supported:
   have a static IP; see [IP Pools](#ip-pools) below for details.
 * `static_ip_pool` - (Optional) A range of IPs permitted to be used as static IPs for
   virtual machines; see [IP Pools](#ip-pools) below for details.
+* `guest_vlan_allowed` - (Optional) True if Network allows guest VLAN tagging.
```
Comment (dataclouder) [inline website/docs/r/network_isolated.html.markdown:65]:
```suggestion
* `guest_vlan_allowed` - (Optional; *v3.11+*) True if Network allows guest VLAN tagging.
```

File: website/docs/r/network_isolated_v2.html.markdown [hunk 10]
```diff
@@ -141,6 +145,7 @@ and inherited from provider configuration)
 * `dns_suffix` - (Optional) A FQDN for the virtual machines on this network
 * `static_ip_pool` - (Optional) A range of IPs permitted to be used as static IPs for
   virtual machines; see [IP Pools](#ip-pools) below for details.
+* `guest_vlan_allowed` - (Optional) True if Network allows guest VLAN tagging.
```
Comment (dataclouder) [inline website/docs/r/network_isolated_v2.html.markdown:148]:
```suggestion
* `guest_vlan_allowed` - (Optional; *v3.11+*) True if Network allows guest VLAN tagging.
```

File: website/docs/r/network_routed.html.markdown [hunk 11]
```diff
@@ -63,6 +65,7 @@ The following arguments are supported:
   have a static IP; see [IP Pools](#ip-pools) below for details.
 * `static_ip_pool` - (Optional) A range of IPs permitted to be used as static IPs for
   virtual machines; see [IP Pools](#ip-pools) below for details.
+* `guest_vlan_allowed` - (Optional) True if Network allows guest VLAN tagging.
```
Comment (dataclouder) [inline website/docs/r/network_routed.html.markdown:68]:
```suggestion
* `guest_vlan_allowed` - (Optional; *v3.11+*) True if Network allows guest VLAN tagging.
```

File: website/docs/r/network_routed_v2.html.markdown [hunk 12]
```diff
@@ -162,6 +166,7 @@ The following arguments are supported:
 * `dns_suffix` - (Optional) A FQDN for the virtual machines on this network
 * `static_ip_pool` - (Optional) A range of IPs permitted to be used as static IPs for
   virtual machines; see [IP Pools](#ip-pools) below for details.
+* `guest_vlan_allowed` - (Optional) True if Network allows guest VLAN tagging.
```
Comment (dataclouder) [inline website/docs/r/network_routed_v2.html.markdown:169]:
```suggestion
* `guest_vlan_allowed` - (Optional; *v3.11+*) True if Network allows guest VLAN tagging.
```

**Comment by Didainius** [comment]
@s-winkelhofer - a few comments/asks:

I have tested the `v2` resources and looks to be working great. 

> @Didainius thank you. We are only using the `_v2` resources. I added the feature to the `non _v2` resources for completeness reasons.

If you don't mind I'd like to keep the original resources (without "_v2") they way they are (without new field). They are using legacy API and we don't intend to maintain them much apart from critical bugs. The reality is that they should be deprecated at some point and then removed later on. They are using legacy XML API, while the `v2` ones use the new OpenAPI endpoints. This was the reason for the introduction of `v2` resources.

I also have a tiny patch that does two things:
* It removes `guest_vlan_allowed = false` in a few places for test configuration (which is the default value anyway because Go boolean becomes `false` by default). This allows for better compatibility check (we don't want to cause any plan issues for people that are upgrading here)
* It adds a few lines to `skip-upgrade-tests.txt` - we're using this for what are called `upgrade` tests. These particular files check if schema or resource changed and it does (because `guest_vlan_allowed` field is added)

You should be able to pick it here https://gist.github.com/Didainius/d7af37d898c8681ef7a9f3b1b487fe66 , save it to a file and then do `git apply filename`. I can't make a PR to your branch directly because the CLAbot we have doesn't like coauthored PRs.

**Comment by Didainius** [comment]
@s-winkelhofer ,
I am working on to remove the field from `v1` resources as we want to pull this one in for 3.11 release.

**Comment by Didainius** [comment]
Closing in favor of #1168

**Pull request closed without merging**